src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenEndpoint.java

```diff
@@ -271,6 +303,19 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	@SuppressWarnings("unchecked")
+	private void exchange(Message<?> message) {
+		Object replyObject = sendAndReceive(message);
+		V reply = null;
+		try {
+			reply = (V) replyObject;
+		}
+		catch (ClassCastException e) {
+			logger.error("Unexpected reply type: " + replyObject.getClass().getName(), e);
+		}
+		this.replyTemplate.send(getMessageBuilderFactory().withPayload(reply).build());
```

Comment:
OTOH, as you said on the call: we can't abuse `messageKey`. According Apache Kafka it isn't necessary unique per message. That key can be used for grouping messages, e.g. per user.
Right, the `key` is used for partition determination as well